### PR TITLE
Homework 4(sort array) done

### PR DESCRIPTION
**Pull request by goharminasyan**

### CODE REVIEW

File: src/algorithms/SortingAlgorithms.java [hunk 1]
```diff
@@ -3,45 +3,136 @@
...
-    public static void sortBySelectionSorting(int[] array){
+    public static void sortBySelectionSorting(int[] array) {
+        for (int k = 0; k < array.length; k++) {
+            for (int l = k + 1; l < array.length; l++) {
+                if (array[l] < array[k]) {
+                    swap(array, k, l);
+                }
+            }
+        }
+    }
```
Comment (Karlen999) [inline src/algorithms/SortingAlgorithms.java:37]: done

File: src/algorithms/SortingAlgorithms.java [hunk 2]
```diff
@@ -3,45 +3,136 @@
...
+                    swap(array,k,l);
+                       int m=k;
+                  while(m>0 && array[m-1]>array[m]){
+                    swap(array,m-1,m);
+                      m--;
+                  }
+                }
+                 k++;
+            }
+        }
```
Comment (Karlen999) [inline src/algorithms/SortingAlgorithms.java:52]: This method does not work correctly

image

File: src/algorithms/SortingAlgorithms.java [hunk 3]
```diff
@@ -3,45 +3,136 @@
...
+            for (int l = k + 1; l < array.length; l++) {
+                if (array[l] < array[k]) {
+                    swap(array,k,l);
+                       int m=k;
+                  while(m>0 && array[m-1]>array[m]){
+                    swap(array,m-1,m);
+                      m--;
+                  }
+                }
+                 k++;
```
Comment (Karlen999) [inline src/algorithms/SortingAlgorithms.java:50]: A for loop stop condition should test the loop counter against an invariant value (i.e. one that is true at both the beginning and ending of every loop iteration). Ideally, this means that the stop condition is set to a local variable just before the loop begins.
Stop conditions that are not invariant are slightly less efficient, as well as being difficult to understand and maintain, and likely lead to the introduction of errors in the future.

File: src/algorithms/SortingAlgorithms.java [hunk 4]
```diff
@@ -3,45 +3,136 @@
...
 
-    public static void sortByMergeSorting(int[] array){
+                while (j < rightArr.length) {
+                    array[k] = rightArr[j];
+                     j++;
+                      k++;
+                }
 
-    }
+            }
```
Comment (Karlen999) [inline src/algorithms/SortingAlgorithms.java:98]: done

File: src/algorithms/SortingAlgorithms.java [hunk 5]
```diff
@@ -3,45 +3,136 @@
...
-    public static void sortByQuickSorting(int[] array){
+            public static void sortByQuickSorting ( int[] array, int begin, int end) {
 
-    }
+                if(begin>=end){return;}
+                int p = partition(array, begin, end);
+       if(begin<p-1){
+            sortByQuickSorting(array, begin, p - 1);}
+       if(p<end){
+            sortByQuickSorting(array, p + 1, end);}
```
Comment (Karlen999) [inline src/algorithms/SortingAlgorithms.java:107]: In order for the code to be readable, you can format the code using **Ctrl+Alt+L**

Comment (goharminasyan) [reply src/algorithms/SortingAlgorithms.java:107]: okay thanks

**Comment by Karlen999** [comment]
reviewed

**Pull request closed without merging**